lib/Dialect/FIRRTL/Transforms/GrandCentral.cpp

```diff
@@ -13,548 +13,1055 @@
+  // Instances link to their definitions via symbols and we don't want to
+  // break this.
   auto builder = OpBuilder::atBlockEnd(circuitOp.getBody());
 
-  // Utility that acts like emitOpError, but does _not_ include a note.  The
-  // note in emitOpError includes the entire op which means the **ENTIRE**
-  // FIRRTL circuit.  This doesn't communicate anything useful to the user other
-  // than flooding their terminal.
-  auto emitCircuitError = [&circuitOp](StringRef message = {}) {
-    return emitError(circuitOp.getLoc(), message);
-  };
+  SymbolTable symbolTable(circuitOp);
 
-  // Examine the Circuit's Annotations doing work to remove Grand Central
-  // Annotations.  Ignore any unprocesssed annotations and rewrite the Circuit's
-  // Annotations with these when done.
-  bool removalError = false;
-  annotations.removeAnnotations([&](auto anno) {
-    if (!anno.isClass("sifive.enterprise.grandcentral.AugmentedBundleType"))
-      return false;
+  CircuitNamespace circuitNamespace(circuitOp);
+  circuitNamespaceRef = &circuitNamespace;
 
-    AugmentedBundleType bundle;
-    if (auto maybeBundle = decodeBundleType(anno))
-      bundle = maybeBundle.getValue();
-    else {
-      emitCircuitError(
-          "'firrtl.circuit' op contained an 'AugmentedBundleType' "
-          "Annotation which did not conform to the expected format")
-              .attachNote()
-          << "the problematic 'AugmentedBundleType' is: '" << anno.getDict()
-          << "'";
+  // Maybe get an "id" from an Annotation.  Generate error messages on the op if
+  // no "id" exists.
+  auto getID = [&](Operation *op,
+                   Annotation annotation) -> Optional<IntegerAttr> {
+    auto id = annotation.getMember<IntegerAttr>("id");
+    if (!id) {
+      op->emitOpError()
+          << "contained a malformed "
+             "'sifive.enterprise.grandcentral.AugmentedGroundType' "
+             "annotation that did not contain an 'id' field";
       removalError = true;
-      return false;
+      return None;
     }
+    return Optional(id);
+  };
 
-    for (auto elt : bundle.elements) {
-      std::pair<StringRef, StringRef> key = {bundle.defName, elt.name};
-      interfaceMap[key] = {elt.tpe, elt.description};
-      interfaceKeys.push_back(key);
+  // Maybe return the lone instance of a module.  Generate errors on the op if
+  // the module is not instantiated or is multiply instantiated.
+  auto exactlyOneInstance = [&](FModuleOp op,
+                                StringRef msg) -> Optional<InstanceOp> {
+    auto instances = symbolTable.getSymbolUses(op, circuitOp);
+    if (!instances || std::distance(instances.getValue().begin(),
+                                    instances.getValue().end()) == 0) {
+      op->emitOpError() << "is marked as a GrandCentral '" << msg
+                        << "', "
+                           "but is never instantiated";
+      return None;
     }
+    if (std::distance(instances.getValue().begin(),
```

Comment:
actually, hoise std::distance up rather than use it in multiple ifs.

Reply:
Cleaned up in the rebase.